### ISSUE TITLE
TypeError: Expected float32, got <larq.constraints.weight_clip object at 0x7fd566f55908> of type 'weight_clip' instead.

### ISSUE THREAD

**Issue by KienPM** [larq/larq#368]
### Describe the bug

When I tried to use QuantConv2D instead of Conv2D from this code https://github.com/qqwweee/keras-yolo3/blob/master/yolo3/model.py, I got error when training
TypeError: Expected float32, got <larq.constraints.weight_clip object at 0x7fd566f55908> of type 'weight_clip' instead.

### To Reproduce

This is my modified code
[code.zip](https://github.com/larq/larq/files/4012066/code.zip)
My optimizer as below
```
case_optimizer = CaseOptimizer(
        (
            Bop.is_binary_variable,  # predicate
            Bop(threshold=1e-6, gamma=1e-3),  # optimizer
        ),
        default_optimizer=Adam(1e-3),
    )
```

### Expected behavior

Any one face this issue? Can you help me please!
### Environment
Ubuntu 16.04
Python 3.6.9

TensorFlow version: tensorflow-gpu 1.15.0
Larq version: larq 0.8.2

**Comment by koenhelwegen** [comment]
Hi, thanks for reporting! Could you post the full error trace and the part of your code were you use larq (in particular how you define your layers)? If you use Bop there is no need to add a `kernel_constraint`, but it shouldn't throw an error either.

**Comment by KienPM** [comment]
Here is my code snippet (you can see the whole code in attached file)
```
@wraps(QuantConv2D)
def DarknetConv2D(*args, **kwargs):
    """Wrapper to set Darknet parameters for Convolution2D."""
    darknet_conv_kwargs = {
        'padding': 'valid' if kwargs.get('strides') == (2, 2) else 'same',
        'input_quantizer': lq.quantizers.ste_sign,
        'kernel_quantizer': lq.quantizers.ste_sign,
        'kernel_constraint': lq.constraints.weight_clip,
        "use_bias": False
    }
    darknet_conv_kwargs.update(kwargs)
    return QuantConv2D(*args, **darknet_conv_kwargs)
```
Here is my traceback
```
Traceback (most recent call last):
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/framework/tensor_util.py", line 324, in _AssertCompatible
    fn(values)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/framework/tensor_util.py", line 263, in inner
    _ = [_check_failed(v) for v in nest.flatten(values)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/framework/tensor_util.py", line 264, in <listcomp>
    if not isinstance(v, expected_types)]
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/framework/tensor_util.py", line 248, in _check_failed
    raise ValueError(v)
ValueError: <larq.constraints.weight_clip object at 0x7f715ce85fd0>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kienpm/dev/keras-yolo3/bnn_train.py", line 226, in <module>
    _main()
  File "/home/kienpm/dev/keras-yolo3/bnn_train.py", line 94, in _main
    callbacks=[logging, checkpoint])
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training.py", line 1296, in fit_generator
    steps_name='steps_per_epoch')
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_generator.py", line 265, in model_iteration
    batch_outs = batch_function(*batch_data)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training.py", line 1016, in train_on_batch
    self._make_train_function()
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training.py", line 2115, in _make_train_function
    params=self._collected_trainable_weights, loss=self.total_loss)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/keras/optimizer_v2/optimizer_v2.py", line 504, in get_updates
    return [self.apply_gradients(grads_and_vars)]
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/larq/optimizers.py", line 130, in apply_gradients
    for optimizer, opt_grads_and_vars in zip(self.optimizers, grad_var_lists)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/larq/optimizers.py", line 130, in <listcomp>
    for optimizer, opt_grads_and_vars in zip(self.optimizers, grad_var_lists)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/keras/optimizer_v2/optimizer_v2.py", line 439, in apply_gradients
    kwargs={"name": name})
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py", line 1940, in merge_call
    return self._merge_call(merge_fn, args, kwargs)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py", line 1947, in _merge_call
    return merge_fn(self._strategy, *args, **kwargs)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/keras/optimizer_v2/optimizer_v2.py", line 483, in _distributed_apply
    var, apply_grad_to_update_var, args=(grad,), group=False))
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py", line 1553, in update
    return self._update(var, fn, args, kwargs, group)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py", line 2165, in _update
    return self._update_non_slot(var, fn, (var,) + tuple(args), kwargs, group)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/distribute/distribute_lib.py", line 2171, in _update_non_slot
    result = fn(*args, **kwargs)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/keras/optimizer_v2/optimizer_v2.py", line 468, in apply_grad_to_update_var
    return var.assign(var.constraint(var))
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/ops/resource_variable_ops.py", line 813, in assign
    value_tensor = ops.convert_to_tensor(value, dtype=self.dtype)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/framework/ops.py", line 1184, in convert_to_tensor
    return convert_to_tensor_v2(value, dtype, preferred_dtype, name)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/framework/ops.py", line 1242, in convert_to_tensor_v2
    as_ref=False)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/framework/ops.py", line 1297, in internal_convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/framework/constant_op.py", line 286, in _constant_tensor_conversion_function
    return constant(v, dtype=dtype, name=name)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/framework/constant_op.py", line 227, in constant
    allow_broadcast=True)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/framework/constant_op.py", line 265, in _constant_impl
    allow_broadcast=allow_broadcast))
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/framework/tensor_util.py", line 449, in make_tensor_proto
    _AssertCompatible(values, dtype)
  File "/home/kienpm/.conda/envs/tf15/lib/python3.6/site-packages/tensorflow_core/python/framework/tensor_util.py", line 331, in _AssertCompatible
    (dtype.name, repr(mismatch), type(mismatch).__name__))
TypeError: Expected float32, got <larq.constraints.weight_clip object at 0x7f715ce85fd0> of type 'weight_clip' instead.
```

**Comment by lgeiger** [comment]
Thanks for the code example. As @koenhelwegen mentioned above since you are using `Bop`, there is no need for a `kernel_quantizer` and `kernel_constraint` since both of these tasks are handled by the optimizer itself.

`lq.constraints.weight_clip` is an alias to `lq.constraints.WeightClip`, similar to how [upstream constraints](https://keras.io/constraints/) work. This means that `lq.constraints.weight_clip` is actually a `class` and therefor needs to be instantiated first (e.g. `lq.constraints.weight_clip()`, or ` `lq.constraints.WeightClip()`).

Unfortunately TensorFlow throws a really confusing error message, so this is not obvious. I fixed our documentation, in #371.
